### PR TITLE
events: Hook on mongo events and log

### PR DESCRIPTION
**Pull request by dsabati**
- Hook to some mongo events and log when event fired.
- Add tests to ensure logs are OK.
- Update mongo-memory-server to v7.6.3 so we can reuse port immediatly after stopping server (cf [mongodb-memory-server issue 330](https://github.com/nodkz/mongodb-memory-server/issues/330))

### CODE REVIEW

File: src/mongo-utils.ts [hunk 1]
```diff
@@ -16,10 +17,39 @@ export class MongoUtils {
...
+			log.info(`-- mongo-utils.ts close --`);
+		});
+		mongoClient.on('serverClosed', () => {
+			log.info(`-- mongo-utils.ts serverClosed --`);
+		});
+		mongoClient.on('reconnect', () => {
+			log.info(`-- mongo-utils.ts reconnect --`);
+		});
+		mongoClient.on('timeout', (e) => {
+			log.warn(`-- mongo-utils.ts timeout --`, e);
```
Comment (BenjD90) [inline src/mongo-utils.ts:39]: Could you format clean log ? that was only an example.

Be carrefull to not add unknown object as context

File: src/mongo-utils.ts [hunk 2]
```diff
@@ -16,10 +17,39 @@ export class MongoUtils {
...
+		});
+
+		const dbClient = await mongoClient.connect();
+		global.dbClient = dbClient;
+		const db = dbClient.db();
 		global.db = db;
+
+		(db as any).s.topology.on('reconnectFailed', () => {
+			log.warn(`-- mongo-utils.ts reconnect failed --`);
```
Comment (BenjD90) [inline src/mongo-utils.ts:48]:
```suggestion
			log.warn(`MongoDB Client reconnect failed.`);
                        process.kill(process.pid, "SIGTERM");
```

if reconnect failed, we have to shut down

Comment (BenjD90) [reply src/mongo-utils.ts:48]: check how is it done in n9-node-routing ;)

**Review by BenjD90 — CHANGES_REQUESTED** [review]

File: src/mongo-utils.ts (hunk 1 above)
Comment (BenjD90) [inline src/mongo-utils.ts:39]: Could you try to test this event ? maybe reduce the allowed timeout to 1ms ?

File: src/mongo-utils.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-console */
```
[the remark this answers is not in the document]
Comment (dsabati) [reply src/mongo-utils.ts]: @BenjD90 done!

File: src/mongo-utils.ts (hunk 1 above)
Comment (dsabati) [reply src/mongo-utils.ts]: @BenjD90 done!

File: src/mongo-utils.ts (hunk 2 above)
Comment (dsabati) [reply src/mongo-utils.ts]: @BenjD90 done!

File: src/mongo-utils.ts [hunk 4]
```diff
@@ -6,28 +6,58 @@ import * as mongodb from 'mongodb';
...
+		const dbClient = await mongoClient.connect();
+		global.dbClient = dbClient;
+		const db = dbClient.db();
 		global.db = db;
-		log.info(`Connected`);
+
+		(db as any).s.topology.on('reconnectFailed', () => {
+			log.warn(`Client reconnection failed.`);
+			if (connectOptions.killProcessOnReconnectFailed) {
+				process.kill(process.pid, 'SIGTERM');
```
Comment (BenjD90) [inline src/mongo-utils.ts:49]: Maybe a warn log telling that we're sending a SIGTERM might be good, no ?

Comment (BenjD90) [inline src/mongo-utils.ts:49]: new idea, have you tried to mock process.kill with mockito ?

File: src/mongo-utils.ts [hunk 5]
```diff
@@ -6,28 +6,58 @@ import * as mongodb from 'mongodb';
...
 
 export class MongoUtils {
 	private static readonly MONGO_ID_REGEXP: RegExp = /^[0-9a-f]{24}$/;
 
 	public static async connect(
 		url: string,
 		options: mongodb.MongoClientOptions = { useNewUrlParser: true },
+		connectOptions: MongoUtilsOptions = { killProcessOnReconnectFailed: true },
 	): Promise<mongodb.Db> {
 		const log = global.log.module('mongo');
```
Comment (BenjD90) [inline src/mongo-utils.ts:19]:
```suggestion
		connectOptions: MongoUtilsOptions = {  },
	): Promise<mongodb.Db> {
		if(connnectOptions.killProcessOnReconnectFailed !== false) {
			connnectOptions.killProcessOnReconnectFailed = true
		}
		const log = global.log.module('mongo');
```

File: src/models/mongo-utils-options.models.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+export interface MongoUtilsOptions {
```
Comment (BenjD90) [inline src/models/mongo-utils-options.models.ts:1]: should be added to the index.ts global export

**Review by BenjD90 — CHANGES_REQUESTED** [review]

File: src/mongo-utils.ts (hunk 4 above)
Comment (dsabati) [reply src/mongo-utils.ts:49]: yes, i tried to mock process.kill using the following code
```typescript
const processMock: NodeJS.Process = Mockito.mock(process);
Mockito.when(processMock.kill(Mockito.anything(), Mockito.anything())).thenReturn();
```
However, I end up with an error saying `processMock.kill is not a function`.
Looking it up online, it seems that the NodeJS.Process cannot be easily mocked...

An alternative way would be to create a new file 'reconnectFailed.js' that runs the code that will call process.kill, and run it in a separate child process...

```typescript
const childProcess = spawn('node', ['reconnectFailed.js']);
await waitFor(2000); // Adjust the time as needed
t.true(childProcess.killed);
```
What's your take on it?

Comment (BenjD90) [reply src/mongo-utils.ts:49]: seems overkill to me, leave it that way, we tried :)

**Pull request merged**